File: 백준/Gold/1687. 행렬 찾기/행렬 찾기.py
import sys

input = sys.stdin.readline


n, m = map(int, input().split())
board = [[0] * (m + 1)] + [[0] + list(map(int, input().strip())) for _ in range(n)]
prefixSum = [[0] * (m + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    for j in range(1, m + 1):
        if board[i][j] == 0:
            board[i][j] = 1
        elif board[i][j] == 1:
            board[i][j] = 0
        prefixSum[i][j] = (
            board[i][j] + prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1]
        )

answer = 0
for col1 in range(1, m + 1):
    for col2 in range(col1, m + 1):
        count = 0
        for row in range(1, n + 1):
            length = (
                prefixSum[row][col2]
                - prefixSum[row][col1 - 1]
                - prefixSum[row - 1][col2]
                + prefixSum[row - 1][col1 - 1]
            )

            if length == col2 - col1 + 1:
                count += length
                answer = max(answer, count)
            else:
                count = 0

print(answer)

File: 백준/Gold/1687. 행렬 찾기/행렬 찾기.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int n, m;
  static int[][] prefixSum;

  public static void main(String[] args) throws IOException {
    setUp();

    int answer = 0;
    for (int col1 = 1; col1 < m + 1; col1++) {
      for (int col2 = col1; col2 < m + 1; col2++) {
        int count = 0;
        for (int row = 1; row < n + 1; row++) {
          if (prefixSum[row][col2] - prefixSum[row][col1 - 1] == col2 - col1 + 1) {
            count += col2 - col1 + 1;
            answer = Math.max(answer, count);
          } else {
            count = 0;
          }
        }
      }
    }

    sb.append(answer);

    output();
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());
    prefixSum = new int[n + 1][m + 1];
    for (int i = 1; i < n + 1; i++) {
      String line = br.readLine();
      for (int j = 1; j < m + 1; j++) {
        prefixSum[i][j] = 1 - (line.charAt(j - 1) - '0');
      }
    }

    for (int i = 1; i < n + 1; i++) {
      for (int j = 1; j < m + 1; j++) {
        prefixSum[i][j] += prefixSum[i][j - 1];
      }
    }
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }


}